### PR TITLE
Fix torch.compile for MoE

### LINKED ISSUE #2029: [Bug] Does Mixtral currently not support torch compile?
### Checklist

- [X] 1. I have searched related issues but cannot get the expected help.
- [X] 2. The bug has not been fixed in the latest version.
- [X] 3. Please note that if the bug-related issue you submitted lacks corresponding environment info and a minimal reproducible demo, it will be challenging for us to reproduce and resolve the issue, reducing the likelihood of receiving feedback.
- [X] 4. If the issue you raised is not a bug but a question, please raise a discussion at https://github.com/sgl-project/sglang/discussions/new/choose Otherwise, it will be closed.
- [X] 5. Please use English, otherwise it will be closed.

### Describe the bug

When I start the Mixtral model with --enable-torch-compile, I get the following error
`
File "/usr/local/lib/python3.10/dist-packages/vllm/model_executor/custom_op.py", line 16, in forward
    return self._forward_method(*args, **kwargs)
TypeError: fused_moe_forward_native() takes from 6 to 8 positional arguments but 9 were given`
I modified the forward method of Fuse_MOE layer and then reported an error
`Exception: Capture cuda graph failed: BatchPrefillWithPagedKVCachePyTorchWrapper::Run(at::Tensor, at::Tensor, std::optional<at::Tensor>, std::optional<at::Tensor>, std::optional<at::Tensor>, at::Tensor, at::Tensor, at::Tensor, bool, unsigned int, bool, int, float, float, float, float, bool)::<lambda()> failed to dispatch data type Float`
I added the parameter --attention-backend triton.The error is now reported.
`[rank0]:E1113 22:38:24.579000 139935494416128 torch/_inductor/runtime/triton_heuristics.py:433] AssertionError: First input (fp32) and second input (bf16) must have the same dtype!`

### Reproduction

CUDA_VISIBLE_DEVICES=0 python3 -m sglang.launch_server --port 6000 --model-path tiny_test_model  --dtype bfloat16 --served-model-name moe-test --host 0.0.0.0 --disable-radix-cache --enable-torch-compile  --attention-backend triton   --torch-compile-max-bs 2

### Environment

`Python: 3.10.15 (main, Sep  7 2024, 18:35:33) [GCC 9.4.0]
CUDA available: True
GPU 0,1,2,3,4,5,6,7: NVIDIA L40
GPU 0,1,2,3,4,5,6,7 Compute Capability: 8.9
CUDA_HOME: /usr/local/cuda
NVCC: Cuda compilation tools, release 12.4, V12.4.131
CUDA Driver Version: 535.104.12
PyTorch: 2.4.0+cu121
flashinfer: 0.1.6+cu124torch2.4
triton: 3.0.0
transformers: 4.46.1
requests: 2.32.3
tqdm: 4.66.6
numpy: 1.26.4
aiohttp: 3.10.10
fastapi: 0.115.4
hf_transfer: 0.1.8
huggingface_hub: 0.26.2
interegular: 0.3.3
packaging: 24.1
PIL: 10.4.0
psutil: 6.1.0
pydantic: 2.9.2
uvicorn: 0.32.0
uvloop: 0.21.0
zmq: 26.2.0
vllm: 0.6.3.post1
multipart: 0.0.17
openai: 1.53.0
anthropic: 0.38.0
NVIDIA Topology: 
        GPU0    GPU1    GPU2    GPU3    GPU4    GPU5    GPU6    GPU7    NIC0    CPU Affinity    NUMA Affinity   GPU NUMA ID
GPU0     X      PIX     PXB     PXB     SYS     SYS     SYS     SYS     PXB     0-31,64-95      0               N/A
GPU1    PIX      X      PXB     PXB     SYS     SYS     SYS     SYS     PXB     0-31,64-95      0               N/A
GPU2    PXB     PXB      X      PXB     SYS     SYS     SYS     SYS     PXB     0-31,64-95      0               N/A
GPU3    PXB     PXB     PXB      X      SYS     SYS     SYS     SYS     PIX     0-31,64-95      0               N/A
GPU4    SYS     SYS     SYS     SYS      X      PIX     PXB     PXB     SYS     32-63,96-127    1               N/A
GPU5    SYS     SYS     SYS     SYS     PIX      X      PXB     PXB     SYS     32-63,96-127    1               N/A
GPU6    SYS     SYS     SYS     SYS     PXB     PXB      X      PXB     SYS     32-63,96-127    1               N/A
GPU7    SYS     SYS     SYS     SYS     PXB     PXB     PXB      X      SYS     32-63,96-127    1               N/A
NIC0    PXB     PXB     PXB     PIX     SYS     SYS     SYS     SYS      X 

Legend:

  X    = Self
  SYS  = Connection traversing PCIe as well as the SMP interconnect between NUMA nodes (e.g., QPI/UPI)
  NODE = Connection traversing PCIe as well as the interconnect between PCIe Host Bridges within a NUMA node
  PHB  = Connection traversing PCIe as well as a PCIe Host Bridge (typically the CPU)
  PXB  = Connection traversing multiple PCIe bridges (without traversing the PCIe Host Bridge)
  PIX  = Connection traversing at most a single PCIe bridge
  NV#  = Connection traversing a bonded set of # NVLinks

NIC Legend:

  NIC0: mlx5_bond_0

ulimit soft: 1048576`

### PR DESCRIPTION
**Pull request by merrymercy**
Fix https://github.com/sgl-project/sglang/issues/2029

### CODE REVIEW

File: python/sglang/srt/layers/fused_moe/patch.py [hunk 1]
```diff
@@ -98,7 +98,9 @@ def fused_moe_forward_native(
     renormalize: bool,
     topk_group: Optional[int] = None,
     num_expert_group: Optional[int] = None,
+    custom_routing_function: Optional[Callable] = None,
 ) -> torch.Tensor:
+    assert custom_routing_function is None
```
Comment (zhyncs) [inline python/sglang/srt/layers/fused_moe/patch.py:103]: why add this?

Comment (merrymercy) [reply python/sglang/srt/layers/fused_moe/patch.py:103]: the new vllm version added this argument

**Review by zhyncs — APPROVED** [review]

**Comment by sitabulaixizawaluduo** [comment]
Do you need triton attention to load the MoE model server?

**Comment by sitabulaixizawaluduo** [comment]
Need triton atttention else?

**Comment by merrymercy** [comment]
No. see usage example here https://github.com/sgl-project/sglang/blob/aae5434bdffd13dc9e00417379168146602553e5/test/srt/test_torch_compile_moe.py#L25

**Pull request merged**